### ISSUE TITLE
SC2015 inconsistent behavior

### ISSUE THREAD

**Issue by rozcietrzewiacz** [koalaman/shellcheck#2976]
#### For bugs
- Rule Id (if any, e.g. SC1000): SC2015
- shellcheck version: online 

#### Here's a snippet or screenshot that shows the problem:

This triggers an error:
```sh
#!/bin/bash
[ "$1" ] && [ "$2" ] || { exit 1; }
```

While this doesn't:
```sh
#!/bin/bash
[ "$1" ] && [ "$2" ] || exit 1
```

#### Here's what I wanted or expected to see:

Consistency.

**Comment by koalaman** [comment]
This inconsistency was intentionally introduced in d603ee1e to reduce the noise of `A && B || C` when C is a single, allowlisted command (echo, printf, exit, return), since these generally aren't a problem. It does not aim to generally determine whether C is a harmless or side effect free command or structure, as this is an infinite rabbit hole.

Somewhat unrelated, the suggestion shouldn't trigger when B is a test, and this is added in 76ff702e9.

**Comment by rozcietrzewiacz** [comment]
@koalaman Oh https://github.com/koalaman/shellcheck/commit/76ff702e9385215a888ed21bf4330f614ab6c355 seems definitely related - it actually solves my main issue 🤩 
Thanks!
